rust/apps/zcash/src/pczt/check.rs

```diff
@@ -0,0 +1,341 @@
+    seed_fingerprint: &[u8; 32],
+    ufvk: &UnifiedFullViewingKey,
+    pczt: &Pczt,
+) -> Result<(), ZcashError> {
+    let xpub = ufvk.transparent().ok_or(ZcashError::InvalidDataError(
+        "transparent xpub is not present".to_string(),
+    ))?;
+    let orchard = ufvk.orchard().ok_or(ZcashError::InvalidDataError(
+        "orchard fvk is not present".to_string(),
+    ))?;
+    check_orchard(&seed_fingerprint, &orchard, &pczt.orchard())?;
+    check_transparent(&seed_fingerprint, &xpub, &pczt.transparent())?;
+    Ok(())
+}
+
+fn check_transparent(
+    seed_fingerprint: &[u8; 32],
+    xpub: &AccountPubKey,
+    bundle: &pczt::transparent::Bundle,
+) -> Result<(), ZcashError> {
+    bundle.inputs().iter().try_for_each(|input| {
+        check_transparent_input(seed_fingerprint, xpub, input)?;
+        Ok(())
+    })?;
+    bundle.outputs().iter().try_for_each(|output| {
+        check_transparent_output(seed_fingerprint, xpub, output)?;
+        Ok(())
+    })?;
+    Ok(())
+}
+
+fn check_transparent_input(
+    seed_fingerprint: &[u8; 32],
+    xpub: &AccountPubKey,
+    input: &pczt::transparent::Input,
+) -> Result<(), ZcashError> {
+    let script = Script(input.script_pubkey().clone());
+    match script.address() {
+        Some(TransparentAddress::PublicKeyHash(hash)) => {
+            let pubkey = input.bip32_derivation().keys().find(|pubkey| {
+                return hash[..] == Ripemd160::digest(Sha256::digest(pubkey))[..];
+            });
+            match pubkey {
+                Some(pubkey) => {
+                    match input.bip32_derivation().get(pubkey) {
+                        Some(bip32_derivation) => {
+                            if seed_fingerprint == &bip32_derivation.seed_fingerprint {
+                                //verify public key
+                                let xpub = xpub.to_inner();
+                                let target = bip32_derivation.derivation_path.iter().try_fold(
+                                    xpub,
+                                    |acc, n| {
+                                        acc.derive_child(ChildNumber::from(*n)).map_err(|_| {
+                                            ZcashError::InvalidPczt(
+                                                "transparent input bip32 derivation path invalid"
+                                                    .to_string(),
+                                            )
+                                        })
+                                    },
+                                )?;
+                                if target.public_key().serialize().to_vec()
+                                    != input.script_pubkey().clone()
+                                {
+                                    return Err(ZcashError::InvalidPczt(
+                                        "transparent input script pubkey mismatch".to_string(),
+                                    ));
+                                }
```

Comment:
I have not confirmed the alterations to your vendored code, but if `target` here has type `bip32::ExtendedPublicKey<secp256k1::PublicKey>` (like it does inside `zcash_primitives::legacy::keys::AccountPubKey`) then there is a bug here: you are comparing the serialized public key to the `script_pubkey`, but for P2PKH scripts the serialized public key appears in the `script_sig`; only its hash is present in `script_pubkey`.

Reply:
This has been fixed.